### DISCUSSION TITLE
Question about behavior of substructmatch function

### DISCUSSION THREAD

**Discussion by gratus907** [rdkit/rdkit#4888] [Q&A] [answered]
Hello!
While using `HasSubstructMatch()` function, I have found an example where results does not match with intuition. 

Consider the following case:
```py
qmol = Chem.MolFromSmarts("cc[N][O-]")
dmol = Chem.MolFromSmiles("O=[N+]([O-])c1ccc(Cl)c([N+](=O)[O-])c1")
dmol.HasSubstructMatch(qmol)
>> True
```
This aligns with intuition - refer to the structure depiction 
![image](https://user-images.githubusercontent.com/20734325/149553727-895766e1-10ba-4209-b6aa-3468ece01ece.png)
We can see that "cc[N][O-]" can be matched, but only when we consider matching N from query to N+ from data as valid match. 

Consider the following case:
```py
qmol = Chem.MolFromSmarts("cc[N][O]")
dmol = Chem.MolFromSmiles("O=[N+]([O-])c1ccc(Cl)c([N+](=O)[O-])c1")
dmol.HasSubstructMatch(qmol)
>> True
```
This example indicates that, matching O (query) to O- (data) is also considered valid. 

However, 
```py
qmol = Chem.MolFromSmarts("cc[N][O-]")
dmol = Chem.MolFromSmiles("O=[N+]([O])c1ccc(Cl)c([N+](=O)[O])c1")
dmol.HasSubstructMatch(qmol)
>> False
```
This indicates that matching O- from query to O in data is invalid. 

I thought this is somewhat inconsistent. (two examples, matching O -> O- and O- -> O must be treated in same way?) Due to insufficient knowledge on chemistry, I could not find appealing reason behind this behavior. Any help is much appreciated!

Few things to note, if helps:
- `dmol = Chem.MolFromSmarts("O=[N+]([O])c1ccc(Cl)c([N+](=O)[O])c1")` gives same results. 
- I am only interested in topologically matching substructures, regardless of whether such molecule can exist or not! So `cc[N][O-]` (using only part of the ring) not being actual molecule doesnt matter and I still think that `HasSubstructMatch` should give consistent results.
- If the behavior is the result of considering hydrogens, the result should be opposite (O- -> O valid, O -> O- is not) but it doesn't seem so.

Thanks!

**Comment by dehaenw** [comment, marked as answer]
hi @gratus907 ,
i think the logic is as follows: if the charge is explicitly not specified in the ~~SMARTS~~ (edit: the mol is instantiated from SMARTS but the search is not a SMARTS query) mol substructure query for a certain any match for that atom is valid, independent of the charge. if the charge is specified, by putting a + or - inside the square brackets, only fragments that have that specific charge at that atom match. you final SMILES does not have a negatively charge oxygen in the nitro-like radical so it will not match. In case you want to search for uncharged N only and exclude -1 or +1, you can specify this in your substructure mol using [N-0].
for example:
```python
qmol = Chem.MolFromSmarts("[N-0]")
dmol = Chem.MolFromSmiles("N")
dmol.HasSubstructMatch(qmol)
>>True
qmol = Chem.MolFromSmarts("[N-0]")
dmol = Chem.MolFromSmiles("[NH4+]")
dmol.HasSubstructMatch(qmol)
>>False
```

**Reply by gratus907** [reply]
Hi @dehaenw 
Thanks for your comment. I understood that by using `[N-0]`, it is possible to force rdkit to match with only neutral-charged atoms. However, this not being the default - as I have demonstrated above, what would I like to describe is "matching criteria being asymmetric" is what bothers me. To summarize, I would like to know why 

```py
qmol = Chem.MolFromSmarts("[N+]")
dmol = Chem.MolFromSmiles("N")
dmol.HasSubstructMatch(qmol)
>> False
```

```py
qmol = Chem.MolFromSmarts("N")
dmol = Chem.MolFromSmiles("[N+]")
dmol.HasSubstructMatch(qmol)
>> True
```
These two differ. Do you have some insight why is this the case?

Again, thanks for answer :)

**Reply by dehaenw** [reply]
Hi @gratus907 ,
i would say there is an inherent asymmetry because one mol is the substructure query and another is the full molecule. You can read about the idea behind it here: http://www.rdkit.org/docs/RDKit_Book.html#atom-atom-matching-in-substructure-queries
i quote:
> The general rule used in the RDKit is that if you don’t specify a property in the query, then it’s not used as part of the matching criteria and that Hs are ignored. This leads to the following behavior

then giving a few examples including some with charged atoms in either query or molecule.

**Reply by gratus907** [reply]
Thanks! The link you gave seems to be what I was looking for :) Apparently I missed what I was looking for in the document pages...
